src/debugger/gdb-stub.c

```diff
@@ -433,7 +448,62 @@ static void _processQSupportedCommand(struct GDBStub* stub, const char* message)
 		}
 		message = end + 1;
 	}
-	strncpy(stub->outgoing, "swbreak+;hwbreak+", GDB_STUB_MAX_LINE - 4);
+	strncpy(stub->outgoing, "swbreak+;hwbreak+;qXfer:features:read+;qXfer:memory-map:read+", GDB_STUB_MAX_LINE - 4);
+}
+
+static void _processQXferCommand(struct GDBStub* stub, const char* params, const char* data) {
+	unsigned offset = 0;
+	unsigned length = 0;
+
+	unsigned index = 0;
+	for (index = 0; params[index] != ','; ++index) {
+		offset <<= 4;
+		offset |= _hex2int(&params[index], 1);
+	}
+
+	++index;
+	unsigned int paramsLength = strlen(params);
+	for (; index + 3 < paramsLength; ++index) {
+		length <<= 4;
+		length |= _hex2int(&params[index], 1);
+	}
+
+	if (length + 1 > GDB_STUB_MAX_LINE) {
+		length = GDB_STUB_MAX_LINE - 1;
+	}
+
+	memset(stub->outgoing, 0, GDB_STUB_MAX_LINE);
+	if (strlen(data) < length + offset) {
+		length = strlen(data) - offset;
+		stub->outgoing[0] = 'l';
+	} else {
+		stub->outgoing[0] = 'm';
+	}
+	strlcpy(&stub->outgoing[1], &data[offset], length + 1);
+}
+
+static void _generateMemoryMapXml(struct GDBStub* stub, char* memoryMap) {
+	size_t index = 0;
+	strncpy(memoryMap, "<memory-map version=\"1.0\">", 27);
+	index += strlen(memoryMap);
+	const struct mCoreMemoryBlock* blocks;
+	size_t nBlocks = stub->d.core->listMemoryBlocks(stub->d.core, &blocks);
+	size_t i;
+	for (i = 0; i < nBlocks; ++i) {
+		if (!(blocks[i].flags & mCORE_MEMORY_MAPPED)) {
+			continue;
+		}
+
+		char* type = blocks[i].flags & mCORE_MEMORY_WRITE ? "ram" : "rom";
+		index += snprintf(&memoryMap[index], GDB_STUB_MAX_LINE - index, "<memory type=\"%s\" start=\"0x%08x\" length=\"0x%08x\"/>", type, blocks[i].start, blocks[i].size);
+	}
+	int amountLeft = GDB_STUB_MAX_LINE - index;
+	if (amountLeft < 14) {
+		strncpy(&memoryMap[index], "</memory-map>", amountLeft);
+	}
+	else {
+		strncpy(&memoryMap[index], "</memory-map>", 14);
+	}
```

Comment:
This conditional is useless.

```suggestion
	strncpy(&memoryMap[index], "</memory-map>", amountLeft);
```